ir/src/transition_constraints/graph.rs

```diff
@@ -159,21 +163,38 @@ impl AlgebraicGraph {
     fn insert_next(
         &mut self,
         symbol_table: &SymbolTable,
-        ident: &str,
+        trace_col_access: &TraceColAccess,
     ) -> Result<(TraceSegment, NodeIndex), SemanticError> {
-        let col_type = symbol_table.get_type(ident)?;
-
-        match col_type {
-            IdentifierType::TraceColumn(column) => {
-                let trace_segment = column.trace_segment();
-                let trace_access = TraceAccess::new(trace_segment, column.col_idx(), 1);
-                let node_index = self.insert_op(Operation::TraceElement(trace_access));
-                Ok((trace_segment, node_index))
+        match trace_col_access {
+            TraceColAccess::Single(ident) => {
+                let col_type = symbol_table.get_type(ident.name())?;
+                if let IdentifierType::TraceColumn(column) = col_type {
+                    let trace_segment = column.trace_segment();
+                    let trace_access = TraceAccess::new(trace_segment, column.col_idx(), 1);
+                    let node_index = self.insert_op(Operation::TraceElement(trace_access));
+                    Ok((trace_segment, node_index))
+                } else {
+                    Err(SemanticError::InvalidUsage(format!(
+                        "Identifier {} was declared as a {} not as a trace column",
+                        ident, col_type
+                    )))
+                }
+            }
+            TraceColAccess::GroupAccess(ident, idx) => {
+                let col_type = symbol_table.get_type(ident.name())?;
+                if let IdentifierType::TraceColumnGroup(column_group) = col_type {
+                    let trace_segment = column_group.trace_segment();
+                    let trace_access =
+                        TraceAccess::new(trace_segment, idx + column_group.start_col_idx(), 1);
+                    let node_index = self.insert_op(Operation::TraceElement(trace_access));
+                    Ok((trace_segment, node_index))
+                } else {
+                    Err(SemanticError::InvalidUsage(format!(
+                        "Identifier {} was declared as a {} not as a trace column group",
+                        ident, col_type
+                    )))
+                }
```

Comment:
There's a lot of duplication here. As mentioned in my review of #101, I think it would be preferable to treat TraceAccess in the same way whether it's a group or a single column. When we're doing things in this graph, we just need a reference within the trace. The only place that really needs to know about the distinction between groups and single columns is the symbol table, because 1) we do validation there, and 2) we may want the column grouping information at the end for codegen.

Reply:
Please let me know if you disagree with the current refactor.